src/client/runner/automation/playback/move.js

```diff
@@ -1,355 +1,364 @@
-                               targetScreenPoint.x ||
-                               (currentCursorPosition.y + styleUtils.getScrollTop(currentDocument)) !==
-                               targetScreenPoint.y;
-
-                    return currentCursorPosition.x !== targetScreenPoint.x ||
-                           currentCursorPosition.y !== targetScreenPoint.y;
-                },
-
-                //moving step
-                function (movingCallback) {
-                    window.setTimeout(function () {
-                        nextStep(movingCallback);
-                    }, 0);
-                },
-
-                //save cursor position
-                function (err) {
-                    if (err)
-                        return;
-                    callback();
-                }
-            );
-        },
+        return whilst(() => !this._isMovingFinished(), () => this._movingStep());
+    }
+
+    _scroll () {
+        var scrollOptions = new OffsetOptions();
 
+        scrollOptions.offsetX = this.offsetX;
+        scrollOptions.offsetY = this.offsetY;
+
+        var scrollAutomation = new ScrollAutomation(this.element, scrollOptions);
+
+        return scrollAutomation.run();
+    }
 
-        callback: function () {
-            actionCallback();
+    _moveToCurrentFrame () {
+        var { x, y }          = Cursor.getPosition();
+        var activeWindow      = Cursor.getActiveWindow();
+        var iframe            = null;
+        var iframeUnderCursor = null;
+        var iframeRectangle   = null;
+
+        var msg = {
+            cmd:       MOVE_REQUEST_CMD,
+            startX:    x,
+            startY:    y,
+            endX:      this.endPoint.x,
+            endY:      this.endPoint.y,
+            modifiers: this.modifiers
+        };
+
+        if (activeWindow.parent === window.self) {
+            iframe            = domUtils.getIframeByWindow(activeWindow);
+            iframeRectangle   = positionUtils.getIFrameClientCoordinates(iframe);
+            iframeUnderCursor = getElementUnderCursor() === iframe;
+
+            msg.left              = iframeRectangle.left;
+            msg.top               = iframeRectangle.top;
+            msg.right             = iframeRectangle.right;
+            msg.bottom            = iframeRectangle.bottom;
+            msg.iframeUnderCursor = iframeUnderCursor;
         }
-    });
-};
+
+        return sendRequestToFrame(msg, MOVE_RESPONSE_CMD, activeWindow)
+            .then(message => {
+                Cursor.setActiveWindow(window);
+
+                if (iframeUnderCursor || window.top !== window.self)
+                    return Cursor.move(message.x, message.y);
+            });
+    }
+
+    run () {
+        return new Promise(resolve => {
+            this
+                ._scroll()
+                .then(() => {
+                    var { x, y } = this._getTargetClientPoint();
```

Comment:
What I don't like about these promise chain is that significant pieces of code comes without any annotation. So it's hard to get what they are actually do. Can we refactor them, so we will have nice Promise chain, like:
```js
doSomething()
.then(() => this._doSomethingElse(args))
.then(() => this._doOneMoreThing())
```